internal/graph/check_fast_path.go

```diff
@@ -0,0 +1,674 @@
+	tk := req.GetTupleKey()
+	objRel := tuple.ToObjectRelationString(tuple.GetType(tk.GetObject()), tk.GetRelation())
+	i, err := checkutil.IteratorReadStartingFromUser(ctx, typesys, ds, req, objRel, nil)
+	if err != nil {
+		return nil, err
+	}
+	iterChan := make(chan *iteratorMsg, 1)
+	if !concurrency.TrySendThroughChannel(ctx, &iteratorMsg{iter: i}, iterChan) {
+		i.Stop() // will not be received to be cleaned up
+	}
+	close(iterChan)
+	return iterChan, nil
+}
+
+func (c *LocalChecker) fastPathComputed(ctx context.Context, req *ResolveCheckRequest, rewrite *openfgav1.Userset) (chan *iteratorMsg, error) {
+	typesys, _ := typesystem.TypesystemFromContext(ctx)
+	tk := req.GetTupleKey()
+	computedRelation := rewrite.GetComputedUserset().GetRelation()
+	rewrittenTupleKey := tuple.NewTupleKey(
+		tk.GetObject(),
+		computedRelation,
+		tk.GetUser(),
+	)
+
+	childRequest := req.clone()
+	childRequest.TupleKey = rewrittenTupleKey
+	objectType := tuple.GetType(tk.GetObject())
+	rel, err := typesys.GetRelation(objectType, computedRelation)
+	if err != nil {
+		return nil, err
+	}
+
+	return c.fastPathRewrite(ctx, childRequest, rel.GetRewrite())
+}
+
+func fastPathUnion(ctx context.Context, iterProducers []*iteratorProducer, iterChan chan *iteratorMsg) {
+	batch := make([]*openfgav1.TupleKey, 0)
+
+	defer func() {
+		// flush
+		if len(batch) > 0 {
+			concurrency.TrySendThroughChannel(ctx, &iteratorMsg{iter: storage.NewStaticTupleKeyIterator(batch)}, iterChan)
+		}
+		close(iterChan)
+		cleanupIteratorProducers(iterProducers)
+	}()
+
+	/*
+		collect iterators from all channels, until all drained
+		start performing union algorithm across the heads, if an iterator is empty, poll once again the producer
+		ask to see if the channel has a new iterator, otherwise consider it done
+	*/
+
+	for len(iterProducers) > 0 {
+		if ctx.Err() != nil {
+			return
+		}
+		var err error
+		iterProducers, err = pollIteratorProducers(ctx, iterProducers)
+		if err != nil {
+			concurrency.TrySendThroughChannel(ctx, &iteratorMsg{err: err}, iterChan)
+			return
+		}
+		allIters := true
+		minKey := ""
+		indexes := make([]int, 0)
+		for idx, producer := range iterProducers {
+			v, err := producer.iter.Head(ctx)
+			if err != nil {
+				if storage.IterIsDoneOrCancelled(err) {
+					producer.iter.Stop()
+					producer.iter = nil
+					allIters = false
+					// we need to ensure we have all iterators at all times
+					break
+				}
+				concurrency.TrySendThroughChannel(ctx, &iteratorMsg{err: err}, iterChan)
+				return
+			}
+			// initialize
+			if idx == 0 {
+				minKey = v.GetObject()
+			}
+
+			if minKey == v.GetObject() {
+				indexes = append(indexes, idx)
+			} else if minKey > v.GetObject() {
+				minKey = v.GetObject()
+				indexes = []int{idx}
+			}
+		}
+
+		if !allIters {
+			// we need to ensure we have all iterators at all times
+			continue
+		}
+
+		// all entries with the same value move forward, but only have to send the value once
```

Comment:
```suggestion
		// all iterators with the same value move forward
```

also I suggest renaming 
- `indexes` to `indexesOfITeratorsWithTheSameObject`
- `minKey` to `minObject`